### ISSUE TITLE
Auto CE Bomb Script

### ISSUE THREAD

**Issue by TheIanator** [Fate-Grand-Automata/FGA#733]
A script to minimize the amount of tedious work involved in creating CE bombs.

I've made a very basic version of such a script for use with AnkuLua and Grand Order Lua. My script is not perfectly optimal QP wise, but I think it strikes a good balance between QP efficiency, amount of time to run, and difficulty of programming. 

I built my script in AnkuLua half because I know the tool better and half because I couldn't manage to get anything to build and function in Android Studio. My code is mess and probably has a bunch of unnecessary waits and hard coded values for my device, but since it barely functions I don't know that it's very likely I'd improve it anytime soon myself. Since now is a good time for CE bomb making, given the increased success rate and half qp cost, I thought I'd post here in case anyone was interested in working it into FGA or simply using it as is. I saw that there were a few closed issues asking for the same thing, but I thought that since I had work to show it might warrant opening a new issue.

What my script currently does:
 
- Selects a level 1 Starvation (1 Star) or Gloom (2 Star) as the base CE
- Scrolls to the bottom of the CE material selection and clicks on every CE starting from the bottom
- Scrolls to the top of the CE material selection and clicks on a single level 1 of CE of the same kind as the base
- Combines them
- Repeats from step 1 until all CEs have been combined together (or more likely it crashes)

What my script assumes you already did:

- Set both filters to show only 1 star OR 2 star CEs
- Set the display scale to the smallest setting
- Locked anything you don't want lost
- Started the script while sitting on an empty "Tap to select a Craft Essence to enhance" screen.

What my script fails to do:

- When it gets close to the end the 2nd copy of the base CE is sometimes already selected and gets deselected instead. This is sometimes a good thing as it insures having a Lv. 1 CE for the next round, but it can also make it take one round too many.
- Stop elegantly. Currently it runs until it doesn't find a Lv. 1 CE and crashes.

To do perfectly optimal CE bombs you should always be using a 1 star CE as a base. My script does not do this, instead consolidating all 1 star CEs together and all 2 stars CEs together. I have a couple reasons for this.

I have never run into a situation where I didn't have enough copies of Starvation or Gloom to consolidate everything. If instead I always used a 1 star as the base I don't think there would be enough copies of Starvation to do everything, requiring programming in more 1 star CEs and a way to make sure that the base and the 2nd copy are the same CE.
Secondly, a 1 star CE base can only hold six 2 star CEs while a 2 star base hand hold 19 (20, if you count the copy of itself), reducing the amount of times the script would need to run to finish.

Additional things that could improve the script:

- The script setting the filters itself rather than assuming their already set. (Open the filter, cancel all, turn on "not selected" and the desired star rarity)
- The script checking if the scale setting is right. (There could be a problem where the region in question is overlapped by the FGA button by default on some systems.)
- Optimizing the 2nd copy selection

The attached zip contains the lua script and the three used pngs.
[Auto CE Bomb.zip](https://github.com/Fate-Grand-Automata/FGA/files/6678422/Auto.CE.Bomb.zip)

**Comment by MathewSachin** [comment]
> What my script currently does:
>- Selects a level 1 Starvation (1 Star) or Gloom (2 Star) as the base CE
>- Scrolls to the bottom of the CE material selection and clicks on every CE starting from the bottom
>- Scrolls to the top of the CE material selection and clicks on a single level 1 of CE of the same kind as the base
>- Combines them
>- Repeats from step 1 until all CEs have been combined together (or more likely it crashes)

I tried it out with Starvation CE. In my case, after it scrolled down and picked a few CEs, it already gained enough exp for level 10. 
So, it couldn't pick another matching CE to ascend and resulted in a Lv 10/10  CE.
I'm not sure, but wouldn't it be better to pick the same CE again first and then pick the other CEs so that we can ascend the CE and hold more levels in case we get a Super/Great success?

Great script otherwise. I could add this in a way that it only launches when on CE enhancement screen. That should ensure that only those who know how the script works would use it because it's not in a general-purpose state right now.

**Comment by MathewSachin** [comment]
Sorry, my previous reply was my mistake.
FGA has a problem with image detection with images that are overly black like in the case of Gloom CE. Overly dark images match with almost anything. I faced this problem because I had selected Gloom as the target instead of Starvation, and it still matched Starvation. At the step where it had to select matching CE, it selected some other CE because of it thinking that every CE is almost the same.

I think I might be able to exploit the way FGA treats black pixels to make it able to detect all 1* CEs with a single image. Same for 2* images. It's the same hack used to detect Lv10 skills without having the image of skills of all servants.

**Comment by MathewSachin** [comment]
I ported the code to FGA: https://github.com/Fate-Grand-Automata/FGA/pull/675/commits/a3f7bdd6896d96fe99d4490fdd50ec9f9eb3c7ee

It's in another branch where I'm redesigning the UI. #675.
You can download the latest APK for that branch from: https://github.com/Fate-Grand-Automata/FGA/actions/runs/952109646

**Comment by the3dsandwich** [comment]
Hey I was testing it out and find it really useful! (Also it's my first time using the jetpack redesign. Have to say it's sooooo pretty and easy to use!)

May I clarify some of the behaviors to make sure it's doing what it's supposed to do?

1. At the beginning, it searches for a matching CE (Gloom or Starvation) that can be at any of the first three rows.
2. Select the same CE (that is also level 1)
3. Click on the bottom of the scroll bar to scroll to the bottom
4. Select all CE from the bottom
5. Confirm

Basically 2. and 3. is reversed so that we don't waste EXPs.

What I observed is that, when I have a lot (300+ I think) of CEs, the base CEs may run out before combining all the 1 star and 2 star CEs. As 3 star CEs pile up the problem goes away on its own, but I wonder if we can add more CEs as base CEs, or even have the script detect the top CE as base CE on its own and start from there.

Also, we had to manually change the ordering from higher rarity to lower rarity (and reverse) for each of the 1 star and 2 star CE. I wonder if it's the intended case or am I thinking it wrong.

Again, I absolutely love this feature! Thanks so much!

**Comment by MathewSachin** [comment]
> Basically 2. and 3. is reversed so that we don't waste EXPs.

It doesn't matter which of 2 or 3 is done first. My earlier comment was wrong. But, I prefer doing 2 first to avoid extra scrolling.

> What I observed is that, when I have a lot (300+ I think) of CEs, the base CEs may run out before combining all the 1 star and 2 star CEs. As 3 star CEs pile up the problem goes away on its own, but I wonder if we can add more CEs as base CEs, or even have the script detect the top CE as base CE on its own and start from there.

I'm already working on adding all the 1* and 2* CEs to the script. I ran out of CEs on the way. I want to save FP till the anniversary when new bronze servants drop on NA. I'll push the change shortly anyways.

> Also, we had to manually change the ordering from higher rarity to lower rarity (and reverse) for each of the 1 star and 2 star CE. I wonder if it's the intended case or am I thinking it wrong.

I'm afraid I didn't understand. Could you elaborate a bit more?
I'm guessing you mean:
- Base CE pickup screen should be filtered to correct rarity and sorted in **_ASCENDING_** order by Level.
- Enhancement material pickup screen should be filtered to correct rarity and sorted in **_DESCENDING_** order by Level.
I'm not certain if these statements I made are correct because I haven't tested much.

**Comment by TheIanator** [comment]
My reasoning for selecting the copy last is that you can select them while at the cap. For example:

Starvation -> Others -> Starvation = 18 Total CEs selected
Starvation -> Starvation -> Others = 17 Total CEs selected

I suppose it becomes a question of whether more time is saved maximizing CE count or minimizing time scrolling.

**Comment by MathewSachin** [comment]
> My reasoning for selecting the copy last is that you can select them while at the cap

Oh, I didn't know about that 😅. I'll revert back the order then.

**Comment by the3dsandwich** [comment]
> I'm afraid I didn't understand. Could you elaborate a bit more?
> I'm guessing you mean:
> - Base CE pickup screen should be filtered to correct rarity and sorted in ASCENDING order by Level.
> - Enhancement material pickup screen should be filtered to correct rarity and sorted in DESCENDING order by Level.

I guess that's the difference. I didn't sort them by level but instead by rarity, and the enhancement material was sorted I think in ascending order. I'll test again with the correct settings later today

I can help test these couple days since we're having a CE enhancement campaign and I can help getting some of the images. Please let me know how I can help!

**Comment by the3dsandwich** [comment]
I made some screenshots of the 1* and 2* CEs if that helps

![Screenshot_1624158262](https://user-images.githubusercontent.com/32169552/122661460-37324e80-d1bd-11eb-9ae5-dbce42f07304.png)
![Screenshot_1624158271](https://user-images.githubusercontent.com/32169552/122661463-3ac5d580-d1bd-11eb-8669-905cc75a71e8.png)
![Screenshot_1624158285](https://user-images.githubusercontent.com/32169552/122661464-3dc0c600-d1bd-11eb-96c4-52626eed63f6.png)
![Screenshot_1624158292](https://user-images.githubusercontent.com/32169552/122661465-41544d00-d1bd-11eb-9d10-83f1fce3b53b.png)
![Screenshot_1624158299](https://user-images.githubusercontent.com/32169552/122661466-43b6a700-d1bd-11eb-862e-acbfaba030c8.png)
![Screenshot_1624158305](https://user-images.githubusercontent.com/32169552/122661467-47e2c480-d1bd-11eb-8f5f-7b92c87f2ca7.png)

**Comment by SirGeorgedon** [comment]
Hi, I am new to this so may I ask how do I add the script into the file that is currently on my phone? Thank you!

**Comment by the3dsandwich** [comment]
If you're asking about testing this CE bomb script **on FGA**, you can download the latest CI build from https://github.com/Fate-Grand-Automata/FGA/actions/runs/952109646

As this script is fairly new, you'll want to be very careful and read through most of the discussions above if not all, and be familiar with installing CI build APKs.

If you're asking about adding the Lua script, then sadly I'm not sure how to do it (and you won't need the Lua script for FGA since we port them over already) (and that's why we need to test them)

**Comment by MathewSachin** [comment]
latest CI build: https://github.com/Fate-Grand-Automata/FGA/actions/runs/952866194

I've added all the 1* and 2* CEs. Also, reverted back the order of picking matching CE.

**Comment by SirGeorgedon** [comment]
Oh nice thank you very much!

**Comment by SirGeorgedon** [comment]
So I ran into a very interesting problem. I got this string here. 

En/combat_0.png

java.io.FileNotFoundException: En/combat_0.png
	at android.content.res.AssetManager.nativeOpenAsset(Native Method)
	at android.content.res.AssetManager.open(AssetManager.java:957)
	at android.content.res.AssetManager.open(AssetManager.java:934)
	at o6.r.f(ImageLoader.kt:3)
	at o6.r.a(ImageLoader.kt:13)
	at t5.f.a(AutoCEBomb.kt:23)
	at p6.b.r(EntryPoint.kt:3)
	at d7.a.run(Thread.kt:1)

Sorry if I am posting at the wrong place.

**Comment by MathewSachin** [comment]
It was accidentally saved as JPG (En/combat_0.jpg) by me 😞

**Comment by the3dsandwich** [comment]
Test run of the latest build: https://www.youtube.com/watch?v=_wgxEI5yZ9k (15:14 for ~500 CEs)

Some notable points:
- The waiting time between scrolling and selecting is quite noticeable, we should be able to reduce it to ~1.5s or even 1s without much problem.
- 20 1* CE won't fill a 2* CE to it's level limit, so when it scrolls back up, it cannot click on another CE of the same kind.
- The above shouldn't be a problem, but when we start with more than 3 rows of the same 2* CE, we will end up with all 3 rows of the CE not being level 1 and the script ends prematurely. (11:58)
- Maybe we can either scroll down to find new base CEs to match, or we can somehow combine the already enhanced CEs to a higher level?

Also if we ever make it so that FGA sets the filters for us, here are some screenshots for TW for matching (if we ever need them)
![Screenshot_1624178061](https://user-images.githubusercontent.com/32169552/122667519-cfdcc480-d1e5-11eb-8d67-78234d37e226.png)
![Screenshot_1624178068](https://user-images.githubusercontent.com/32169552/122667525-d2d7b500-d1e5-11eb-9094-17c6137203dd.png)

**Comment by MathewSachin** [comment]
> The waiting time between scrolling and selecting is quite noticeable, we should be able to reduce it to ~1.5s or even 1s without much problem.

👍 I'll make that change.

> 20 1* CE won't fill a 2* CE to it's level limit, so when it scrolls back up, it cannot click on another CE of the same kind.

I think we're only meant to feed 2* or higher CE into 2* CE, i.e. same rarity or higher CEs into themselves. Again, I'm not used to this method of making CE bombs, so it would be great if @TheIanator could confirm this.

> The above shouldn't be a problem, but when we start with more than 3 rows of the same 2* CE, we will end up with all 3 rows of the CE not being level 1 and the script ends prematurely. (11:58)

Personally, I'm fine with this as it is a good time to step in and combine all the enhanced CEs. Or maybe we could scroll down.

> Maybe we can either scroll down to find new base CEs to match, or we can somehow combine the already enhanced CEs to a higher level?

Scrolling down is a valid option, we probably would have to limit how much we scroll in that case (maybe 5 times).

Combining the already enhanced CEs doesn't look possible. We're only able to detect Level 1 CEs and nothing else.
We match images of Level 1 CEs, e.g. ![image](https://user-images.githubusercontent.com/10654537/122668303-0827d700-d1d5-11eb-9dfc-a54607da5a8f.png).
We technically could do that by cropping out the level part from the image, but there's always a risk of overcapping and wasting EXP.

> Also if we ever make it so that FGA sets the filters for us, here are some screenshots for TW for matching (if we ever need them)

I think we could (whenever that is) blindly click on filters without matching because they seem to be in the same locations as NA.

**Comment by SirGeorgedon** [comment]
> It was accidentally saved as JPG (En/combat_0.jpg) by me 😞

HAHA! So I will wait for you to code an updated version first then?

**Comment by TheIanator** [comment]
> > 20 1* CE won't fill a 2* CE to it's level limit, so when it scrolls back up, it cannot click on another CE of the same kind.
> 
> I think we're only meant to feed 2* or higher CE into 2* CE, i.e. same rarity or higher CEs into themselves. Again, I'm not used to this method of making CE bombs, so it would be great if @TheIanator could confirm this.

That's how I've been doing it, it's not really the right way of doing it, but I think it's good enough and there is a lot of conflicting information about what the actual correct way is.

**Comment by SirGeorgedon** [comment]
Did you guys change the code yourself to get the CE leveling thing working?

**Comment by the3dsandwich** [comment]
@SirGeorgedon Latest build is working. No need to change the code. While I did pull the code and changed some files but it's for another feature and I didn't change the behavior here. Check this and see if the problem is still there. https://github.com/Fate-Grand-Automata/FGA/actions/runs/954139205

**Comment by SirGeorgedon** [comment]
Yup it doesn't show the error right now but for some reason it keeps saying No suitable target CE found. Should I put the image here?

**Comment by aradovn** [comment]
I read this topic for awhile and still don't know how to get this making CE bomb start. I installed lastest CI build 1371 but don't see the choice related to make CE bomb.

**Comment by the3dsandwich** [comment]
> I read this topic for awhile and still don't know how to get this making CE bomb start. I installed lastest CI build 1371 but don't see the choice related to make CE bomb.

You start the script from the CE Enhancement screen, leave all selections blank. 

Remember to select proper filters and sorting options before starting the script, which is:

- if you will combine 2* CEs: filter base CE by 2*, sorted by decreasing level. Filter enhancement material CE by 2*, sorted by increasing level
-

**Comment by the3dsandwich** [comment]
Are you on JP by any chance? This is a fairly new feature so we should expect some quirks and the JP UI change might be a breaking point. (Plus some other parts may not behave as expected, we're still finding the best way to use this script)

It would be nice if you can provide some image and we can see where our problems are.

**Comment by the3dsandwich** [comment]
Oh yeah by the way, there's a quirk when selecting enhancement CEs, it'll tap on the stop button when the overlay was at the default location (bottom left), since it taps on CEs one by one from the bottom right.

It's not a problem if we put the button to the top right corner, but it's just a quirk people need to look out for.

**Comment by aradovn** [comment]
> > I read this topic for awhile and still don't know how to get this making CE bomb start. I installed lastest CI build 1371 but don't see the choice related to make CE bomb.
> 
> You start the script from the CE Enhancement screen, leave all selections blank.
> 
> Remember to select proper filters and sorting options before starting the script, which is:
> 
> * if you will combine 2* CEs: filter base CE by 2*, sorted by INCREASING level. Filter enhancement material CE by 2*, sorted by DECREASING level
> * if you will combine 1* CEs: filter base CE by 1*, sorted by INCREASING level. Filter enhancement material CE by 1*, sorted by DECREASING level
> 
> Keep in mind that this script only combines level 1 base CEs so you'll have to step in once you have a bunch of ~lvl10 CEs.
> 
> We're also aware that there are many ways of making a CE bomb and this is just one way of doing it.

Yay, it's running now, thank you very much! 
At first I got the same error as SirGeorgedon, so I sort the CE to 4x4 grid like then start to change position of FGA overlay button. Like you said the top right corner work well. And after that even when I put it back at original bottom left it still work.

**Comment by Draw99Gray** [comment]
The easiest and most widespread method of making CE bombs is to just pick a 1* CE and MLB it, then stuffing it full up to lv50 starting with all 3* trash CEs and down
It costs a bit more QP than the supposed CE bomb method but it's weeeell worth in the "keeping your sanity" part and would be way easier to script
you could put a disclaimer that the script will eat all unlocked CEs, but since CE bomb making is usually started at high level i doubt you'll get complaints

**Comment by MathewSachin** [comment]
> The easiest and most widespread method of making CE bombs is to just pick a 1* CE and MLB it, then stuffing it full up to lv50 starting with all 3* trash CEs and down

Yup, this is just what I do manually.


> would be way easier to script

No, I honestly doubt it would be easier without text recognition. But, I'd be very happy if you could code up a prototype. 😄

**Comment by Draw99Gray** [comment]
the way i see it, it would be "picking all CEs until there are no more OR max level is reached" and since when max level is reached the CE just deselects itself it's a good stopping/restarting point
correct me if i'm wrong

**Comment by d7c0** [comment]
Is there a way of making this work with CE EXP Cards?
